File: backend/src/content/cleaner.rs
//! HTML Content Cleaning Service
//!
//! Ported from TypeScript cleaner.ts - removes scripts, styles, navigation,
//! iframes, and Kortext/Prescribed Reading boilerplate.

use lol_html::{element, rewrite_str, RewriteStrSettings};

/// Kortext phrases - containers with these phrases are removed
const KORTEXT_PHRASES: &[&str] = &[
    "Sign in to Kortext",
    "Open book in new window",
    "You will only be able to access the book on Kortext",
    "kortext.com",
    "launchReader",
    "emailKortextSupport",
];

/// Prescribed Reading phrases
const PRESCRIBED_READING_PHRASES: &[&str] = &[
    "Prescribed Reading",
];

/// Container classes to remove when they contain unwanted text
const CONTAINER_CLASSES: &[&str] = &[
    "no-overflow",
    "box",
    "generalbox", 
    "prescribed-reading",
];

/// Clean HTML content
pub fn clean_html_content(html: &str) -> String {
    clean_html_with_token(html, None)
}

/// Clean HTML content and optionally fix image URLs with token
pub fn clean_html_with_token(html: &str, token: Option<&str>) -> String {
    if html.is_empty() {
        return String::new();
    }

    let original_len = html.len();
    
    // First pass: streaming removal of definitely unwanted elements
    let result = rewrite_str(
        html,
        RewriteStrSettings {
            element_content_handlers: vec![
                // Remove script elements
                element!("script", |el| {
                    el.remove();
                    Ok(())
                }),
                // Remove style elements
                element!("style", |el| {
                    el.remove();
                    Ok(())
                }),
                // Remove link stylesheets
                element!("link[rel='stylesheet']", |el| {
                    el.remove();
                    Ok(())
                }),
                element!("link[rel=\"stylesheet\"]", |el| {
                    el.remove();
                    Ok(())
                }),
                // Remove iframes (Kortext embeds, etc.)
                element!("iframe", |el| {
                    el.remove();
                    Ok(())
                }),
                // Remove nav elements
                element!("nav", |el| {
                    el.remove();
                    Ok(())
                }),
                // Remove elements with navigation classes
                element!(".navigation", |el| {
                    el.remove();
                    Ok(())
                }),
                element!(".breadcrumb", |el| {
                    el.remove();
                    Ok(())
                }),
                element!("#page-header", |el| {
                    el.remove();
                    Ok(())
                }),
                element!(".modified", |el| {
                    el.remove();
                    Ok(())
                }),
                element!(".activity-navigation", |el| {
                    el.remove();
                    Ok(())
                }),
            ],
            ..RewriteStrSettings::default()
        },
    );

    let mut output = match result {
        Ok(s) => s,
        Err(e) => {
            tracing::error!("lol_html rewrite failed: {}", e);
            return html.to_string();
        }
    };

    // Second pass: remove unwanted containers using scraper
    output = remove_unwanted_containers(&output);
    
    // Remove duplicate headings
    output = remove_duplicate_headings(&output);
    
    // Post-process: clean images
    output = clean_images(&output);
    
    // Post-process: remove empty paragraphs
    output = remove_empty_paragraphs(&output);
    
    // Fix image URLs with token if provided
    if let Some(token) = token {
        output = fix_image_urls(&output, token);
    }
    
    // Fix entity encoding
    output = output.replace("&amp;nbsp;", " ");
    output = output.replace("&nbsp;", " ");
    output = output.replace("&amp;amp;", "&");
    
    tracing::debug!(
        "Cleaned HTML: {} -> {} bytes",
        original_len,
        output.len()
    );

    output
}

/// Remove containers that have unwanted Kortext/Prescribed Reading content
fn remove_unwanted_containers(html: &str) -> String {
    use scraper::{Html, Selector};
    
    let mut output = html.to_string();
    
    // Combine all unwanted phrases
    let all_phrases: Vec<&str> = KORTEXT_PHRASES.iter()
        .chain(PRESCRIBED_READING_PHRASES.iter())
        .copied()
        .collect();
    
    // For each container class, check if it contains unwanted text and remove it
    for class in CONTAINER_CLASSES {
        let selector_str = format!(".{}", class);
        let selector = match Selector::parse(&selector_str) {
            Ok(s) => s,
            Err(_) => continue,
        };
        
        // Re-parse after each class to get fresh DOM
        let document = Html::parse_document(&output);
        
        for element in document.select(&selector) {
            // Get the full HTML including attributes (to catch onclick="launchReader()")
            let element_html = element.html();
            let element_html_lower = element_html.to_lowercase();
            
            // Check if this container has any unwanted phrase in text OR attributes
            let has_unwanted = all_phrases.iter().any(|phrase| 
                element_html_lower.contains(&phrase.to_lowercase())
            );
            
            if has_unwanted {
                output = output.replace(&element_html, "");
                tracing::debug!("Removed container .{} with Kortext/Prescribed Reading content", class);
            }
        }
    }
    
    output
}

/// Remove duplicate headings (keeps first occurrence, removes subsequent identical ones)
fn remove_duplicate_headings(html: &str) -> String {
    use scraper::{Html, Selector};
    use std::collections::HashSet;
    
    let document = Html::parse_document(html);
    let mut output = html.to_string();
    let mut seen_headings: HashSet<String> = HashSet::new();
    
    // Check h2 and h3 headings (most common duplicates)
    for tag in ["h2", "h3"] {
        let selector = match Selector::parse(tag) {
            Ok(s) => s,
            Err(_) => continue,
        };
        
        for element in document.select(&selector) {
            let text: String = element.text().collect::<String>().trim().to_string();
            
            // Skip empty headings
            if text.is_empty() {
                continue;
            }
            
            let key = format!("{}:{}", tag, text);
            
            if seen_headings.contains(&key) {
                // This is a duplicate - remove it
                let element_html = element.html();
                output = output.replacen(&element_html, "", 1);
                tracing::debug!("Removed duplicate heading: {}", text);
            } else {
                seen_headings.insert(key);
            }
        }
    }
    
    output
}

/// Clean images - remove spacers and data URIs
fn clean_images(html: &str) -> String {
    use scraper::{Html, Selector};
    
    let document = Html::parse_document(html);
    let mut output = html.to_string();
    
    if let Ok(selector) = Selector::parse("img") {
        for element in document.select(&selector) {
            if let Some(src) = element.value().attr("src") {
                // Remove spacer images
                if src.is_empty() || src.starts_with("data:image/gif;base64") || src.contains("spacer") {
                    output = output.replace(&element.html(), "");
                }
            } else {
                // No src, remove
                output = output.replace(&element.html(), "");
            }
        }
    }
    
    output
}

/// Remove empty paragraphs
fn remove_empty_paragraphs(html: &str) -> String {
    if let Ok(re) = regex::Regex::new(r"<p[^>]*>\s*(&nbsp;|\s)*\s*</p>") {
        re.replace_all(html, "").to_string()
    } else {
        html.to_string()
    }
}

/// Fix image URLs to include authentication token
fn fix_image_urls(html: &str, token: &str) -> String {
    let mut result = html.to_string();
    
    if let Ok(re) = regex::Regex::new(r#"(<img[^>]+src=["'])([^"']+)(["'])"#) {
        result = re.replace_all(&result, |caps: &regex::Captures| {
            let prefix = &caps[1];
            let url = &caps[2];
            let suffix = &caps[3];
            
            // Skip if already has token
            if url.contains("token=") {
                return caps[0].to_string();
            }
            
            // Skip data URIs
            if url.starts_with("data:") {
                return caps[0].to_string();
            }
            
            // Skip external URLs (not Moodle)
            if url.starts_with("http") && !url.contains("mylms.vossie.net") {
                return caps[0].to_string();
            }
            
            // Add token to Moodle URLs or relative URLs
            let separator = if url.contains('?') { "&" } else { "?" };
            format!("{}{}{}token={}{}", prefix, url, separator, token, suffix)
        }).to_string();
    }
    
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_removes_scripts() {
        let html = "<html><body><p>Before</p><script>alert('x');</script><p>After</p></body></html>";
        let cleaned = clean_html_content(html);
        assert!(!cleaned.contains("<script>"));
        assert!(cleaned.contains("Before"));
        assert!(cleaned.contains("After"));
    }

    #[test]
    fn test_removes_kortext_container() {
        let html = r#"<html><body><p>Content</p><div class="no-overflow">Sign in to Kortext stuff</div><p>More</p></body></html>"#;
        let cleaned = clean_html_content(html);
        assert!(!cleaned.contains("Kortext"));
        assert!(cleaned.contains("Content"));
        assert!(cleaned.contains("More"));
    }
    
    #[test]
    fn test_removes_container_with_launchreader() {
        let html = r##"<html><body><p>Content</p><div class="no-overflow"><button onclick="launchReader()">Open</button></div><p>More</p></body></html>"##;
        let cleaned = clean_html_content(html);
        assert!(!cleaned.contains("launchReader"));
        assert!(cleaned.contains("Content"));
        assert!(cleaned.contains("More"));
    }

    #[test]
    fn test_removes_iframes() {
        let html = r#"<html><body><p>Content</p><iframe src="https://kortext.com/embed"></iframe><p>More</p></body></html>"#;
        let cleaned = clean_html_content(html);
        assert!(!cleaned.contains("<iframe"));
        assert!(cleaned.contains("Content"));
    }
    
    #[test]
    fn test_preserves_normal_content() {
        let html = "<html><body><h2>1. Learning outcomes</h2><p>By the end of this chapter, you should be able to understand calculus.</p></body></html>";
        let cleaned = clean_html_content(html);
        assert!(cleaned.contains("Learning outcomes"));
        assert!(cleaned.contains("By the end of this chapter"));
        assert!(cleaned.contains("understand calculus"));
    }
    
    #[test]
    fn test_preserves_box_without_kortext() {
        let html = r#"<html><body><div class="box"><p>Important regular content here</p></div></body></html>"#;
        let cleaned = clean_html_content(html);
        assert!(cleaned.contains("Important regular content"));
    }
    
    #[test]
    fn test_removes_prescribed_reading_container() {
        let html = r#"<html><body><p>Content</p><div class="no-overflow">Prescribed Reading for this module</div><p>More</p></body></html>"#;
        let cleaned = clean_html_content(html);
        assert!(!cleaned.contains("Prescribed Reading"));
        assert!(cleaned.contains("Content"));
        assert!(cleaned.contains("More"));
    }
}

File: mylms-dashboard/src/lib/content/cleaner.ts
/**
 * HTML Content Cleaning Service
 * 
 * This service handles cleaning of raw Moodle HTML content to create a
 * distraction-free reading experience. It removes navigation, tracking,
 * and unwanted third-party integrations (like Kortext).
 */

import { MOODLE_CONFIG } from '@/config/moodle';

/**
 * Main function to clean HTML content from Moodle
 * 
 * @param html Raw HTML content from Moodle
 * @returns Cleaned HTML suitable for display
 */
export async function cleanHtmlContent(html: string): Promise<string> {
  if (!html) return '';
  
  try {
    const cheerio = await import('cheerio');
    const $ = cheerio.load(html);

    // Remove scripts and styles
    removeScriptsAndStyles($);
    
    // Remove navigation and metadata
    removeNavigationElements($);
    
    // Remove Kortext and Prescribed Reading junk
    removeUnwantedContent($);
    
    // Remove broken images
    cleanImages($);
    
    // Remove empty paragraphs
    removeEmptyParagraphs($);

    let result = $.html();
    
    // Fix common entity encoding issues
    // Replace &nbsp; with normal space
    result = result.replace(/&nbsp;/g, ' ');
    // Fix double encoded &nbsp;
    result = result.replace(/&amp;nbsp;/g, ' ');
    // Fix double encoded ampersands (if they are just text)
    // We have to be careful not to break HTML entities, but &amp;amp; -> &amp; is usually what we want for text
    result = result.replace(/&amp;amp;/g, '&amp;');
    
    console.log(`[Clean] Input length: ${html.length}, Output length: ${result.length}`);
    return result;

  } catch (error) {
    console.error('[Clean] Error cleaning HTML:', error);
    return html; // Fallback to original HTML if cleaning fails
  }
}

/**
 * Remove all scripts, styles, and stylesheet links
 */
function removeScriptsAndStyles($: any): void {
  $('script').remove();
  $('style').remove();
  $('link[rel="stylesheet"]').remove();
}

/**
 * Remove navigation elements and metadata
 */
function removeNavigationElements($: any): void {
  MOODLE_CONFIG.UNWANTED_CONTENT.NAVIGATION_SELECTORS.forEach(selector => {
    $(selector).remove();
  });
}

/**
 * Remove Kortext iframes and Prescribed Reading blocks
 */
function removeUnwantedContent($: any): void {
  // Remove all iframes (usually external tools like Kortext)
  $('iframe').remove();
  
  // Collect all unwanted text patterns
  const unwantedTexts = [
    ...MOODLE_CONFIG.UNWANTED_CONTENT.KORTEXT_PHRASES,
    ...MOODLE_CONFIG.UNWANTED_CONTENT.PRESCRIBED_READING_PHRASES
  ];

  unwantedTexts.forEach(text => {
    // Find all elements containing the text
    $(`:contains('${text}')`).each((_i: number, el: any) => {
      // Check if this element has children that also contain the text
      // If it does, it's a parent container, so we skip it (we'll catch the child)
      // If it doesn't, it's likely the leaf or near-leaf node we want to remove
      const hasChildWithText = $(el).children().filter((_j: number, child: any) => 
        $(child).text().includes(text)
      ).length > 0;
      
      if (!hasChildWithText) {
        // This is the deepest element with the text.
        // Now we want to remove its container if it's a "box" or "section"
        const containerSelector = MOODLE_CONFIG.UNWANTED_CONTENT.CONTAINER_SELECTORS.join(', ');
        const container = $(el).closest(containerSelector);
        
        if (container.length) {
          // Remove the container if it's one of these classes
          container.remove();
        } else {
          // Otherwise just remove the element itself (and maybe its parent if it's a p tag)
          const parent = $(el).parent();
          if (parent.is('p') || parent.is('div.text_to_html')) {
            parent.remove();
          } else {
            $(el).remove();
          }
        }
      }
    });
  });
}

/**
 * Clean up images - remove spacers, icons, and images in unwanted containers
 */
function cleanImages($: any): void {
  $('img').each((_i: number, el: any) => {
    const src = $(el).attr('src');
    
    // Remove spacer images or icons
    if (!src || src.startsWith('data:image/gif;base64') || src.includes('icon')) {
      $(el).remove();
      return;
    }
    
    // Remove images inside the Prescribed Reading box if it wasn't caught above
    if ($(el).closest('.prescribed-reading').length) {
      $(el).remove();
    }
  });
}

/**
 * Remove empty paragraphs that have no content or media
 */
function removeEmptyParagraphs($: any): void {
  $('p').each((_i: number, el: any) => {
    // Only remove if truly empty (no text, no images, no iframes)
    if ($(el).text().trim() === '' && $(el).find('img, iframe').length === 0) {
      $(el).remove();
    }
  });
}

/**
 * Download an HTML file from Moodle with token authentication
 * 
 * @param fileUrl The URL of the HTML file to download
 * @param token Moodle authentication token
 * @returns The HTML content, or null if download failed
 */
export async function downloadHtmlFile(fileUrl: string, token: string): Promise<string | null> {
  try {
    // Add token to URL (same as download-itcpb-html-fixed.js)
    let downloadUrl = fileUrl;
    if (!fileUrl.includes('token=') && !fileUrl.includes('wstoken=')) {
      const separator = fileUrl.includes('?') ? '&' : '?';
      downloadUrl = `${fileUrl}${separator}token=${token}`;
    }

    console.log(`[File] Downloading from: ${fileUrl.substring(0, 80)}...`);

    const response = await fetch(downloadUrl);
    
    if (!response.ok) {
      console.error(`[File] HTTP error: ${response.status}`);
      return null;
    }

    const html = await response.text();

    // Verify not an error response
    if (html.includes('"errorcode":"missingparam"') || 
        (html.trim().startsWith('{') && html.includes('"error"'))) {
      console.error('[File] Got error response');
      return null;
    }

    return html;
  } catch (error) {
    console.error('[File] Download failed:', error);
    return null;
  }
}
